### PR TITLE
testlapack: replace isOrthogonal with residualOrthogonal and remove unused helpers

### PR DESCRIPTION
**Pull request by vladimir-ch**
This one's a bit bigger but the split into commits is quite reasonable and hopefully helps.

Please take a look.

### CODE REVIEW

File: lapack/testlapack/dgeqp3.go [hunk 1]
```diff
@@ -20,120 +22,95 @@ type Dgeqp3er interface {
...
+			case some:
+				// Some columns are free, some are leading columns.
+				jpvt[j] = rnd.Intn(2) - 1 // -1 or 0
+				name += "free=some"
+			case none:
+				// All columns are leading.
+				jpvt[j] = 0
+				name += "free=none"
+			default:
+				panic("bad freedom")
```
Comment (kortschak) [inline lapack/testlapack/dgeqp3.go]: :smile:

File: lapack/testlapack/dgeqp3.go [hunk 2]
```diff
@@ -20,120 +22,95 @@ type Dgeqp3er interface {
...
-			if !equalApproxGeneral(got, want, 1e-13) {
-				t.Errorf("Case %v,  Q*R != A*P\nQ*R=%v\nA*P=%v", c, got, want)
+		}
+		// Allocate a slice for scalar factors of elementary
+		// reflectors and fill it with random numbers. Dgeqp3
+		// will overwrite them with valid data.
+		k := min(m, n)
+		tau := make([]float64, k)
+		for i := range tau {
+			tau[i] = rnd.Float64()
```
Comment (kortschak) [inline lapack/testlapack/dgeqp3.go]: Just a thought, I think pretty much all of our random tests use `rnd.Float64()` values, which means they are in [0,1). Is that a problem?

File: lapack/testlapack/dggsvd3.go [hunk 3]
```diff
@@ -20,6 +20,8 @@ type Dggsvd3er interface {
 }
 
 func Dggsvd3Test(t *testing.T, impl Dggsvd3er) {
+	const tol = 1e-14
```
Comment (kortschak) [inline lapack/testlapack/dggsvd3.go]: This is too strict for the netlib tests; 1e-13 works though (this was the value at the call site previously - maybe `10*tol`?).

```
$ CGO_LDFLAGS="-L/usr/local/lib -lopenblas" go test
--- FAIL: TestDggsvd3 (0.00s)
    dggsvd3.go:172: test 8: Vᵀ*B*Q != D2*[ 0 R ]
        Vᵀ*B*Q:
        {Rows:10 Cols:10 Data:[0.010482112652853516 -0.0014056207038094135 0.0006559519382838101 0.003725590442949931 0.006925541668426766 0.007073327941066766 0.001930960930201993 -0.0030724510258088927 0.0010703926989341687 0.004112105844484928 1.6653345369377348e-16 -1.4898783041686252 -0.1218855891352658 0.3659739628713561 -0.06290878558098531 -0.05993028176504687 -0.016933473811997113 -0.33656036701520964 0.2832008131783849 -0.3640892786201968 -3.0531133177191805e-16 -2.3592239273284576e-16 -1.0208307461361052 0.5546638903436011 -0.1898505835556743 0.7034316833818617 -0.6217149180221678 0.15111680339476768 0.9580993428884317 -0.06862912295989301 2.220446049250313e-16 3.0531133177191805e-16 -1.6653345369377348e-16 -1.8611296253425205 -0.8361701088012632 -0.4229783536274635 -0.8427208215286233 0.4962942208067193 -1.3217653092399257 0.33709390093039016 -8.881784197001252e-16 -9.159339953157541e-16 -9.992007221626409e-16 9.08995101411847e-16 -2.5781301435876354 0.7847959492276326 -0.1360207566183565 -1.0513358146275495 1.2654109103132873 -0.24968139531023686 3.3306690738754696e-16 4.163336342344337e-17 4.163336342344337e-17 -1.0408340855860843e-16 -1.1102230246251565e-16 2.0784331505443 0.09418010018954537 0.2410211197213878 0.9072359064092793 0.9981334069018213 -2.914335439641036e-16 3.8163916471489756e-16 -5.204170427930421e-16 -4.527628272299467e-16 -3.7470027081099033e-16 -9.08995101411847e-16 -2.134071471113827 -0.31047041864183567 -0.4874817233896455 -0.853051588315365 -1.3600232051658168e-15 2.914335439641036e-16 3.469446951953614e-16 7.684824998577255e-16 1.942890293094024e-16 -4.614364446098307e-16 3.0531133177191805e-16 2.9840841226386066 2.2084106453754493 -1.1368512271223956 1.4849232954361469e-15 6.175615574477433e-16 5.967448757360216e-16 -3.8077180297690916e-16 4.996003610813204e-16 5.082739784612045e-16 9.853229343548264e-16 -1.1657341758564144e-15 -3.820044002534826 -1.2412821905696736 -2.220446049250313e-16 0 1.2212453270876722e-15 -7.216449660063518e-16 1.3322676295501878e-15 -1.3877787807814457e-15 -1.1102230246251565e-16 3.3306690738754696e-15 -2.220446049250313e-16 -5.57454084980766] Stride:10}
        D2*[ 0 R ]:
        {Rows:10 Cols:10 Data:[0.010482112652853641 -0.001405620703809334 0.0006559519382838444 0.0037255904429495564 0.006925541668426938 0.0070733279410669465 0.0019309609302016915 -0.0030724510258089885 0.0010703926989342393 0.004112105844485231 0 -1.489878304168625 -0.12188558913525951 0.3659739628713477 -0.0629087855809863 -0.05993028176505228 -0.01693347381199522 -0.3365603670152095 0.28320081317837387 -0.36408927862019547 0 0 -1.020830746136105 0.5546638903436163 -0.18985058355566675 0.7034316833818655 -0.6217149180221601 0.15111680339476397 0.9580993428884422 -0.06862912295989353 0 0 0 -1.8611296253425214 -0.8361701088012634 -0.4229783536274645 -0.8427208215286233 0.4962942208067193 -1.3217653092399266 0.33709390093039016 0 0 0 0 -2.5781301435876376 0.7847959492276324 -0.1360207566183568 -1.0513358146275502 1.2654109103132876 -0.24968139531023548 0 0 0 0 0 2.0784331505442997 0.09418010018954472 0.2410211197213882 0.9072359064092795 0.998133406901822 0 0 0 0 0 0 -2.134071471113827 -0.310470418641836 -0.48748172338964574 -0.8530515883153639 0 0 0 0 0 0 0 2.9840841226386075 2.2084106453754497 -1.1368512271223943 0 0 0 0 0 0 0 0 -3.8200440025348263 -1.2412821905696745 0 0 0 0 0 0 0 0 0 -5.574540849807666] Stride:10}
FAIL
```

File: lapack/testlapack/general.go [hunk 4]
```diff
@@ -1498,3 +1498,157 @@ func svdJobString(job lapack.SVDJob) string {
...
+
+	// Compute
+	//  work = work - Q * Qᵀ = I - Q * Qᵀ
+	// or
+	//  work = work - Qᵀ * Q = I - Qᵀ * Q
+	blas64.Syrk(transq, -1, q, 1, work)
+	return dlansy(lapack.MaxColumnSum, blas.Upper, work.N, work.Data, work.Stride)
+}
+
+func dlansy(norm lapack.MatrixNorm, uplo blas.Uplo, n int, a []float64, lda int) float64 {
```
Comment (kortschak) [inline lapack/testlapack/general.go]: `// dlansy is a local implementation of Dlansy to keep code paths independent.`

File: lapack/testlapack/general.go [hunk 5]
```diff
@@ -1498,3 +1498,157 @@ func svdJobString(job lapack.SVDJob) string {
...
+			}
+		}
+		return max
+	default:
+		// lapack.Frobenius:
+		panic("not implemented")
+	}
+}
+
+func dlange(norm lapack.MatrixNorm, m, n int, a []float64, lda int) float64 {
```
Comment (kortschak) [inline lapack/testlapack/general.go]: `// dlange is a local implementation of Dlange to keep code paths independent.`

File: lapack/testlapack/dlauu2.go [hunk 6]
```diff
@@ -87,24 +88,24 @@ func dlauuTest(t *testing.T, dlauu func(blas.Uplo, int, []float64, int), uplo bl
...
+			r := cloneGeneral(blas64.General{Rows: n, Cols: n, Data: a, Stride: lda})
 			if uplo == blas.Upper {
 				// Use aCopy as a dense representation of the upper triangular U.
 				u := aCopy
 				ldu := lda
-				// Compute U * Uᵀ and store the result into want.
+				// Compute U*Uᵀ - A and store the result into R.
 				bi.Dgemm(blas.NoTrans, blas.Trans, n, n, n,
-					1, u, ldu, u, ldu, 0, want, ldwant)
+					1, u, ldu, u, ldu, -1, r.Data, r.Stride)
```
Comment (kortschak) [inline lapack/testlapack/dlauu2.go]: Nice.

File: lapack/testlapack/dgeqp3.go (hunk 1 above)
Comment (vladimir-ch) [reply lapack/testlapack/dgeqp3.go]: I think you are the author of this :-)

File: lapack/testlapack/dgeqp3.go (hunk 2 above)
Comment (vladimir-ch) [reply lapack/testlapack/dgeqp3.go]: In tests that I write I use NormFloat64 but in general I don't think it's a problem, it fulfils the purpose of testing.

File: lapack/testlapack/dggsvd3.go (hunk 3 above)
Comment (vladimir-ch) [reply lapack/testlapack/dggsvd3.go]: Changed to `1e-13`.

File: lapack/testlapack/dlauu2.go (hunk 6 above)
Comment (vladimir-ch) [reply lapack/testlapack/dlauu2.go]: Fortrans's TESTING directory is a useful source of learning material.

File: lapack/testlapack/general.go (hunk 4 above)
Comment (vladimir-ch) [reply lapack/testlapack/general.go]: Thanks, done.

File: lapack/testlapack/general.go (hunk 5 above)
Comment (vladimir-ch) [reply lapack/testlapack/general.go]: Thanks, done.

**Comment by vladimir-ch** [comment]
PTAL

**Pull request merged**